### PR TITLE
update harmony to new implementation

### LINKED ISSUE #299: Switch harmony to harmony-torch
Switch the harmony implementation to avoid Liscence

### PR DESCRIPTION
**Pull request by Intron7**
Fixes #299

### CODE REVIEW

**Comment by flying-sheep** [comment]
I’ll check this out tomorrow, it’s too big to start now!

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+            objectives_harmony,
+            max_iter_clustering,
+            sigma,
+            block_proportion,
+        )
+
+        Z_hat = _correction(Z, R, Phi, O, ridge_lambda, correction_method)
+        Z_norm = _normalize_cp(Z_hat, p=2)
+        if verbose:
+            print(f"\tCompleted {i + 1} / {max_iter_harmony} iteration(s).")
```
Comment (flying-sheep) [inline src/rapids_singlecell/preprocessing/_harmony/__init__.py:213]: don’t you have some logging infra you should use instead?

I don’t want to see a single `print` statement in any library I use (if it has a CLI, that one may have print statements)

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import cupy as cp
+import numpy as np
+from cuml import KMeans as cumlKMeans
```
Comment (flying-sheep) [inline src/rapids_singlecell/preprocessing/_harmony/__init__.py:7]:
```suggestion
from cuml import KMeans as CumlKMeans
```

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+            sigma,
+            block_proportion,
+        )
+
+        Z_hat = _correction(Z, R, Phi, O, ridge_lambda, correction_method)
+        Z_norm = _normalize_cp(Z_hat, p=2)
+        if verbose:
+            print(f"\tCompleted {i + 1} / {max_iter_harmony} iteration(s).")
+
+        if _is_convergent_harmony(objectives_harmony, tol=tol_harmony):
```
Comment (flying-sheep) [inline src/rapids_singlecell/preprocessing/_harmony/__init__.py:215]: so this one is an out parameter of `_clustering`? what else is being modified? You should make that clear

**Review by flying-sheep — CHANGES_REQUESTED** [review]
The `harmonize` function has a really nice layout. it’s easy to follow what it does, nice!

However it seems like you’re establishing a bunch of parameters that get reused and never changed after the initialization. You could e.g.

- create it a frozen dataclass with methods so you can use the parameters using `self.<name>`, or
- create a `NamedTuple` that you can pass around containing the parameters

File: src/rapids_singlecell/preprocessing/_harmony/_kernels/_normalize.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+normalize_kernel_optimized = r"""
+({0} * X, int rows, int cols) {
+    __shared__ {0}  shared[32];  // Shared memory for partial sums (one per thread)
+
+    int row = blockIdx.x;  // One block per row
+    int tid = threadIdx.x;  // Thread index within the block
+
+    // Ensure we're within matrix bounds
+    if (row >= rows) return;
```
[the remark this answers is not in the document]
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/_kernels/_normalize.py:13]: Yes kinda. That come from blocks that are overlapping. eg. 32 but only 29 cells.

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+    block_dim = 32
+    grid_dim = rows  # One block per row
+
+    normalize_p1 = _get_normalize_kernel_optimized(X.dtype)
+    # Launch the kernel
+    normalize_p1((grid_dim,), (block_dim,), (X, rows, cols))
+    return X
+
+
+def _normalize_cp(X: cp.ndarray, p: int = 2) -> cp.ndarray:
```
[the remark this answers is not in the document]
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:40]: Thats the name of the variable in torch

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+    E = cp.dot(Pr_b, cp.sum(R, axis=0, keepdims=True))
+    O = cp.dot(Phi.T, R)
+
+    objectives_harmony = []
+    _compute_objective(Y_norm, Z_norm, R, theta, sigma, O, E, objectives_harmony)
+
+    return R, E, O, objectives_harmony
+
+
+def _clustering(
```
[the remark this answers is not in the document]
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:253]: this is a special kind fuzzy clustering.

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+        batch_vec = batch_mat[batch_key]
+
+    else:
+        df = batch_mat[batch_key].astype("str")
+        batch_vec = df.apply(lambda row: ",".join(row), axis=1)
+
+    return batch_vec.astype("category")
+
+
+def _one_hot_tensor_cp(X: pd.Series) -> cp.array:
```
[the remark this answers is not in the document]
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:64]: added a docstring

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+        Weight of the entropy term in objective function.
+
+    block_proportion
+        Proportion of block size in one update operation of clustering step.
+
+    theta
+        Weight of the diversity penalty term in objective function.
+
+    tau
+        Discounting factor on ``theta``. By default, there is no discounting.
```
[the remark this answers is not in the document]
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:134]: Thoses are the names of in the original implementation so I'm not changing it

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+    if n_clusters is None:
+        n_clusters = int(min(100, n_cells / 30))
+    theta = (cp.ones(n_batches) * theta).astype(Z.dtype)
+
+    if tau > 0:
+        theta = theta * (1 - cp.exp(-N_b / (n_clusters * tau)) ** 2)
+
+    theta = theta.reshape(1, -1)
+    assert block_proportion > 0 and block_proportion <= 1
+    assert correction_method in {"fast", "original"}
```
[the remark this answers is not in the document]
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:176]: done

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+
+    cp.random.seed(random_state)
+
+    # Initialize centroids
```
[the remark this answers is not in the document]
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py]: Done

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (flying-sheep) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:44]: But nut the name we usually use. We use `axis` right?

File: src/rapids_singlecell/preprocessing/_harmony/_fuses.py
[the remark this answers is not in the document]
Comment (flying-sheep) [reply src/rapids_singlecell/preprocessing/_harmony/_fuses.py:1]: they still have weird names with capital letters in them

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (flying-sheep) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:275]: more details please, in a comment or docstring.

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py (hunk 3 above)
Comment (flying-sheep) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:224]: this is still unanswered

**Review by flying-sheep — CHANGES_REQUESTED** [review]
Please also check the review summary: https://github.com/scverse/rapids_singlecell/pull/308#pullrequestreview-2508439874

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+def _initialize_centroids(
+    Z_norm: cp.ndarray,
+    n_clusters: int,
+    sigma: float,
+    Pr_b: cp.ndarray,
+    Phi: cp.ndarray,
+    *,
+    theta: cp.ndarray,
+    random_state: int = 0,
+    n_init: int = 10,
```
[the remark this answers is not in the document]
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:244]: removed

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:275]: Added a better docstring

File: src/rapids_singlecell/preprocessing/_harmony/__init__.py (hunk 3 above)
Comment (Intron7) [reply src/rapids_singlecell/preprocessing/_harmony/__init__.py:223]: I added a docstring to both _is_convergent functions

**Review by flying-sheep — APPROVED** [review]
You said you’re going to do a follow-up PR that moves all the various reused parameters (Phi, R, E, O, theta, …) into a container. Otherwise looks good!

**Pull request merged**